pkg/virt-controller/watch/volume-migration/volume-migration.go

```diff
@@ -0,0 +1,327 @@
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+
+	k8sv1 "k8s.io/api/core/v1"
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+	"kubevirt.io/kubevirt/pkg/controller"
+	storagetypes "kubevirt.io/kubevirt/pkg/storage/types"
+)
+
+const invalidUpdateErrMsg = "The volume can only be reverted to the previous version during the update"
+
+// invalidVols includes the invalid volumes for the volume migration
+type invalidVols struct {
+	hotplugged  []string
+	fs          []string
+	shareable   []string
+	luns        []string
+	dvs         []string
+	dvTemplates []string
+}
+
+func (vols *invalidVols) errorMessage() string {
+	var s strings.Builder
+	if len(vols.hotplugged) < 0 && len(vols.fs) < 0 &&
+		len(vols.shareable) < 0 && len(vols.luns) < 0 {
+		return ""
+	}
+	s.WriteString("invalid volumes to update with migration:")
+	if len(vols.hotplugged) > 0 {
+		s.WriteString(fmt.Sprintf(" hotplugged: %v", vols.hotplugged))
+	}
+	if len(vols.fs) > 0 {
+		s.WriteString(fmt.Sprintf(" filesystems: %v", vols.fs))
+	}
+	if len(vols.shareable) > 0 {
+		s.WriteString(fmt.Sprintf(" shareable: %v", vols.shareable))
+	}
+	if len(vols.luns) > 0 {
+		s.WriteString(fmt.Sprintf(" luns: %v", vols.luns))
+	}
+	if len(vols.dvs) > 0 {
+		s.WriteString(fmt.Sprintf(" DVs: %v", vols.dvs))
+	}
+	if len(vols.dvTemplates) > 0 {
+		s.WriteString(fmt.Sprintf(" DV templates: %v", vols.dvTemplates))
+	}
+
+	return s.String()
+}
+
+func updatedVolumesMapping(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) map[string]string {
+	updateVols := make(map[string]string)
+	vmVols := make(map[string]string)
+	// New volumes
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		if name := storagetypes.PVCNameFromVirtVolume(&v); name != "" {
+			vmVols[v.Name] = name
+		}
+	}
+	// Old volumes
+	for _, v := range vmi.Spec.Volumes {
+		name := storagetypes.PVCNameFromVirtVolume(&v)
+		if claim, ok := vmVols[v.Name]; ok && name != claim {
+			updateVols[v.Name] = claim
+		}
+	}
+	return updateVols
+}
+
+func ValidateVolumes(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) (bool, string) {
+	var invalidVols invalidVols
+	if vmi == nil {
+		return false, "cannot validate the migrated volumes for an empty VMI"
+	}
+	if vm == nil {
+		return false, "cannot validate the migrated volumes for an empty VM"
+	}
+	updatedVols := updatedVolumesMapping(vmi, vm)
+	valid := true
+	disks := storagetypes.GetDisksByName(&vmi.Spec)
+	filesystems := storagetypes.GetFilesystemsFromVolumes(vmi)
+	dvTemplates := make(map[string]bool)
+	for _, t := range vm.Spec.DataVolumeTemplates {
+		dvTemplates[t.Name] = true
+	}
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		_, ok := updatedVols[v.Name]
+		if !ok {
+			continue
+		}
+
+		// Datavolumes should be transformed into PVCs
+		if v.VolumeSource.DataVolume != nil {
```

Comment:
@mhenriks this is to ensure that we don't use any DVs in the update VM. As we discussed last week, I will do a follow-up PR to remove this restriction. However, in this PR I would like to avoid it

Reply:
Just to be clear, my point is that pvc, standalone datavolume, or datavolume template may be PersistentVolumeClaim volume type or DataVolume volume type.  Basically there is no meaningful difference between DataVolume and PersistentVolumeClaim KubeVirt volume types.  My understanding is that DataVolumeTemplate handling will come in a follow up but I'm not sure how that is applicable here

Reply:
I'm not sure I understand the comment. For standalone DVs or DV templates, kubevirt has the DataVolume set in the `VolumeSource`, for PVC it is the PersistentVolumeClaim.
This current version allows only the DV to be migrated to PVC. The final VM should only have PVCs no DVs

Reply:
> For standalone DVs or DV templates, kubevirt has the DataVolume set in the VolumeSource

Not always!  This is a legit VM definition:

```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: testvm
spec:
  dataVolumeTemplates:
  - metadata:
      name: testvm-fedora
    spec:
      storage:
        resources:
          requests:
            storage: 7Gi
      source:
        registry:
          url: "docker://quay.io/kubevirtci/fedora-with-test-tooling:fc35-2208011009"
          pullMethod: node
  running: true
  template:
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: datavolumedisk1
        resources:
          requests:
            memory: 256M
      volumes:
      - persistentVolumeClaim:
          claimName: testvm-fedora
        name: datavolumedisk1
```

Reply:
This is legit too:

```
apiVersion: cdi.kubevirt.io/v1beta1
kind: VolumeImportSource
metadata:
  name: vis
spec:
  source:
    registry:
      url: "docker://quay.io/kubevirtci/fedora-with-test-tooling:fc35-2208011009"
      pullMethod: node
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: testvm-fedora
spec:
  dataSourceRef:
    apiGroup: cdi.kubevirt.io
    kind: VolumeImportSource
    name: vis
  resources:
    requests:
      storage: 7Gi
  storageClassName: ocs-storagecluster-ceph-rbd
  volumeMode: Block
  accessModes:
  - ReadWriteMany
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    foo: bar
  name: testvm
spec:
  running: true
  template:
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: datavolumedisk1
        resources:
          requests:
            memory: 256M
      volumes:
      - dataVolume:
          name: testvm-fedora
        name: datavolumedisk1
```

Reply:
But still this version only supports the volume migration to PVCs not DVs so in the end you don't need to have any migrated volumes which corrispond to a DV. I understand that we don't want to keep this in the long terms. But this assumption avoids me the need to consider all these combination.
I'll handle this in a separate PR

Reply:
Maybe in the end it will be easier to just include DVs, but for now, I'd rather just consider the simplest case and add more combination later

Reply:
So for my example above with populators even though DataVolumes are not used (just the volume type) this PR will not migrate the volume?

Reply:
One thing. This check is done on the new VM api. This implies that the new version cannot have DVs or DV templates but the original one can.

Reply:
To me you are conflating the DataVolume resource with the DataVolume volume type.  I see no reason to not get the volume type handling correct in this PR.  I don't think it's that hard.  Just see if there's a DV of the given name exists.  If so mark it invalid

Reply:
Could be, dvs have caused me some headache. I'll review this more carefully

Reply:
I just removed the restriction of the DVs, now we can also migrate to a DV. I also added a new functional test to cover this case